[feature](bitmap) Support in bitmap syntax and bitmap runtime filter

luozenglin opened the pull request:
1. Support in bitmap syntax,like 'where k1 in (select bitmap_column from tbl)';
2. Support bitmap runtime filter. Generate a bitmap filter using the right table bitmap and push it down to the left table storage layer for filtering.

# Proposed changes

Issue Number: close #14338

Performance：
- before :
select name, count(name) as num from tbl where id in (select bc from bitmap_table lateral view explode_bitmap(bitmap_column) tmp as bc where k1 = 1001) and name != '' group by name order by num desc limit 10;

- in bitmap + bitmap filter : 
select name, count(name) as num from tbl c where c.id in (select bitmap_column from bitmap_table where k1 = 1001) group by name order by num desc limit 10;

Enable bloom filter and bitmap filter, and use the default parameters:
||tbl 100w, bitmap 100w|tbl 1000w, bitmap 100w|tbl 2000w, bitmap 200w| tbl 10000w,bitmap200w|
|:--|:--|:--|:--|:--|
|before|0.31s|0.54s|3.78s|27.49s|
|After optimization|0.06s|0.31s|0.91s|6.35s|

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

BiteTheDDDDt commented:
Why the test of `before` and `bitmap filter` have diffrent sql?

luozenglin commented:
> Why the test of `before` and `bitmap filter` have diffrent sql?

Because the `bitmap_column` column in the `bitmap_table` table is bitmap type, this pr supports the new syntax of `in bitmap`, thus avoiding the use of table fuction.

yangzhg commented on be/src/vec/exprs/vbitmap_predicate.h:
```diff
@@ -0,0 +1,64 @@
+
+#pragma once
+
+#include "vexpr.h"
```
```suggestion
#include "vec/exprs/vexpr.h"
```

yangzhg commented on be/src/vec/exprs/vbitmap_predicate.cpp:
```diff
@@ -0,0 +1,109 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#include "vbitmap_predicate.h"
```
```suggestion
#include "vec/exprs/vbitmap_predicate.h"
```

yangzhg reviewed (commented):
Add some docs to describe how to use bitmap runtime filter please

luozenglin commented:
> Add some docs to describe how to use bitmap runtime filter please

done

BiteTheDDDDt commented on be/src/olap/bitmap_filter_predicate.cpp:
```diff
@@ -0,0 +1,46 @@
+    M(TYPE_BIGINT)
+
+namespace doris {
+ColumnPredicate* BitmapFilterColumnPredicateFactory::create_column_predicate(
```
Better move it to `predicate_creator.h` and do some absract with `create_column_predicate`

luozenglin replied:
done.

BiteTheDDDDt commented on be/src/olap/bitmap_filter_predicate.h:
```diff
@@ -0,0 +1,133 @@
+            : ColumnPredicate(column_id),
+              _filter(filter),
+              _specific_filter(static_cast<SpecificFilter*>(_filter.get())),
+              _be_exec_version(be_exec_version) {}
```
`_be_exec_version` seem unused

luozenglin replied:
I have removed it. thanks.

BiteTheDDDDt commented on be/src/vec/exprs/vbitmap_predicate.cpp:
```diff
@@ -0,0 +1,109 @@
+        argument_template.emplace_back(std::move(column), child->data_type(), child->expr_name());
+    }
+
+    _be_exec_version = state->be_exec_version();
```
same

BiteTheDDDDt commented on be/src/exprs/runtime_filter.cpp:
```diff
@@ -1868,6 +1913,25 @@ Status RuntimePredicateWrapper::get_push_vexprs(std::vector<doris::vectorized::V
+        node.__set_opcode(TExprOpcode::RT_FILTER);
+        node.__isset.vector_opcode = true;
+        node.__set_vector_opcode(to_in_opcode(_column_return_type));
+        node.__set_is_nullable(vprob_expr->root()->is_nullable());
```
Maybe we can set is_nullable to false

luozenglin replied:
done.

cambyzju commented on be/src/exprs/bitmapfilter_predicate.h:
```diff
@@ -0,0 +1,149 @@
+    virtual bool find(uint64 data) const = 0;
+    virtual bool is_empty() = 0;
+    virtual Status assign(BitmapValue* bitmap_value) = 0;
+    virtual Status merge(BitmapFilterFuncBase* bitmap_func, ObjectPool* pool) = 0;
```
why not just remove `merge` function?

luozenglin replied:
removing it directly is ok

cambyzju commented on be/src/olap/bitmap_filter_predicate.h:
```diff
@@ -0,0 +1,121 @@
+        return new_size;
+    }
+
+    std::string _debug_string() override {
```
it is better to make `debug_string` public,  support debug outside `BitmapFilterColumnPredicate`.

luozenglin replied:
There is a public `debug_string()` in the base class of `BitmapFilterColumnPredicate`, and `debug_string()` calls `_debug_string()`

cambyzju reviewed (dismissed):
LGTM

luozenglin committed:
[feature](bitmap) Support in bitmap syntax and bitmap runtime filter

1.Support in bitmap syntax, like 'where k1 in (select bitmap_column from tbl)';
2.Support bitmap runtime filter. Generate a bitmap filter using the right table bitmap and push it down to the left table storage layer for filtering.

yangzhg reviewed (approved):
LGTM
